python/pyiceberg/expressions/__init__.py

```diff
@@ -317,7 +317,7 @@ def __eq__(self, other):
         return self.term == other.term if isinstance(other, UnboundPredicate) else False
 
     @abstractmethod
-    def bind(self, schema: Schema, case_sensitive: bool = True) -> BoundPredicate:
+    def bind(self, schema: Schema, case_sensitive: bool = True) -> BooleanExpression:
```

Comment:
@Fokko, this is more accurate because binding can produce constants like `AlwaysTrue()`, but I'm not sure that we want to do it if it messes up type inference after binding.

Reply:
I think that's more accurate as well 👍🏻 We'll lose the bound type indeed, but for me, that's okay. Probably we also want to annotate the `__new__` methods with type annotations.

Reply:
I had to add a type parameter for the bound result to `Unbound` to get types working, but it didn't require many changes.